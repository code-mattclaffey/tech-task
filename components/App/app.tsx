import React, { useState } from "react";
import { LocationsContext } from "../LocationsContext/LocationsContext";

type ProviderProps = {};

export const App: React.FunctionComponent<ProviderProps> = ({
    children,
    ...rest
}) => {
    const [state, setState] = useState({
        selectedLocations: [],
        ...rest
    });

    const removeSelectedLocation = (locationName: string) => {
        const { selectedLocations } = state;
        const filteredSelectedLocations = selectedLocations.filter(
            ({ location }) => location !== locationName
        );

        setState({
            ...state,
            selectedLocations: filteredSelectedLocations
        });
    };

    const setSelectedLocations = (results: Array<any>) => {
        const { selectedLocations } = state;
        const newSelectedLocations: any = [...selectedLocations, ...results];
        setState({ 
            ...state,
            selectedLocations: newSelectedLocations,
        });
    };

    const data = {
        ...state,
        removeSelectedLocation,
        setSelectedLocations
    };

    return (
        // @ts-ignore
        <LocationsContext.Provider value={data}>
            {children}
        </LocationsContext.Provider>
    );
};
