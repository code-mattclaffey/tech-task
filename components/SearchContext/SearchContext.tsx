import React, { useState, useContext } from "react";
import { getLocations, getLatestParametersFromCity } from "../../apis/api";
import { LocationsContext } from '../LocationsContext/LocationsContext';


export const SearchContext = React.createContext({
    searchTerm: "",
    searchResults: [],
    setSearchTerm: (searchTerm: string) => searchTerm,
    handleKeyEvent: (event: any) => event,
    resetSearchResults: () => null,
    selectLocation: (city: string) => city
});

type ProviderProps = {
    searchResults?: Array<any>;
};

export const SearchContextProvider: React.FunctionComponent<ProviderProps> = ({
    children,
    searchResults,
}) => {
    const { setSelectedLocations } = useContext(LocationsContext);

    const [state, setState] = useState({
        searchTerm: "",
        searchResults: searchResults || []
    });

    const reduceDulplicateCities = (allCities: any, currentCity: any) => {
        const { city } = currentCity;

        const cityAlreadyExistsInObject = allCities.filter((cityObj: any) => {
            return cityObj.city === city;
        });

        if (cityAlreadyExistsInObject.length === 0) {
            allCities.push(currentCity);
        }

        return allCities;
    };

    const setSearchTerm = (searchTerm: string) => {
        setState({ ...state, searchTerm });
    };

    const handleKeyEvent = () => {
        const { searchTerm } = state;
        if (searchTerm.length === 0) {
            setState({ ...state, searchResults: [] });
            return;
        }

        getLocations(searchTerm).then(({ results }) => {
            const searchResults = results.reduce(reduceDulplicateCities, []);
            setState({ ...state, searchResults });
        });
    };

    const resetSearchResults = () => {
        setState({ ...state, searchResults: [] });
    };

    const selectLocation = (city: string) => {
        getLatestParametersFromCity(city).then(({ results }) => {
            setSelectedLocations(results);
            setState({ ...state, searchResults: [] });
        });
    };

    const data = {
        ...state,
        setSearchTerm,
        handleKeyEvent: handleKeyEvent,
        resetSearchResults,
        selectLocation
    };

    return (
        // @ts-ignore
        <SearchContext.Provider value={data}>
            {children}
        </SearchContext.Provider>
    );
};
