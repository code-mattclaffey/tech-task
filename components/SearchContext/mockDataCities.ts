

export const mockData = {"meta":{"name":"openaq-api","license":"CC BY 4.0","website":"https://docs.openaq.org/","page":1,"limit":100,"found":7},"results":[{"id":"GB-28","country":"GB","city":"Manchester","cities":["Manchester"],"location":"Manchester Piccadilly","locations":["Manchester Piccadilly"],"sourceName":"DEFRA","sourceNames":["DEFRA"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-2.23788,"latitude":53.48152},"firstUpdated":"2016-02-27T21:00:00.000Z","lastUpdated":"2020-01-18T13:00:00.000Z","parameters":["no2","o3","pm10","pm25","so2"],"countsByMeasurement":[{"parameter":"no2","count":26469},{"parameter":"o3","count":28442},{"parameter":"pm10","count":265},{"parameter":"pm25","count":27504},{"parameter":"so2","count":27380}],"count":110060},{"id":"GB-152","country":"GB","city":"Manchester","cities":["Manchester"],"location":"Manchester Sharston","locations":["Manchester Sharston"],"sourceName":"DEFRA","sourceNames":["DEFRA"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-2.23922,"latitude":53.37131},"firstUpdated":"2016-05-04T15:00:00.000Z","lastUpdated":"2020-01-18T13:00:00.000Z","parameters":["no2","o3"],"countsByMeasurement":[{"parameter":"no2","count":27470},{"parameter":"o3","count":27447}],"count":54917},{"id":"GB-137","country":"GB","city":"Manchester","cities":["Manchester"],"location":"Salford Eccles","locations":["Salford Eccles"],"sourceName":"DEFRA","sourceNames":["DEFRA"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-2.33414,"latitude":53.48481},"firstUpdated":"2016-02-27T21:00:00.000Z","lastUpdated":"2020-01-18T13:00:00.000Z","parameters":["no2","pm10","pm25"],"countsByMeasurement":[{"parameter":"no2","count":27921},{"parameter":"pm10","count":27574},{"parameter":"pm25","count":27696}],"count":83191},{"id":"US-1977","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Miller State Park","locations":["Miller State Park"],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.87862,"latitude":42.86183},"firstUpdated":"2017-01-18T17:00:00.000Z","lastUpdated":"2020-01-18T12:00:00.000Z","parameters":["co","o3","pm10","pm25","so2"],"countsByMeasurement":[{"parameter":"co","count":19253},{"parameter":"o3","count":18861},{"parameter":"pm10","count":19489},{"parameter":"pm25","count":19573},{"parameter":"so2","count":18403}],"count":95579},{"id":"US-815","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Miller State Park","locations":["Miller State Park"],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.8789,"latitude":42.8619},"firstUpdated":"2016-03-06T19:00:00.000Z","lastUpdated":"2017-01-18T16:00:00.000Z","parameters":["co","o3","pm10","pm25","so2"],"countsByMeasurement":[{"parameter":"co","count":7141},{"parameter":"o3","count":7170},{"parameter":"pm10","count":3114},{"parameter":"pm25","count":7168},{"parameter":"so2","count":7063}],"count":31656},{"id":"US-1673","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Nashua - Gilson Rd.","locations":["Nashua - Gilson Rd."],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.445,"latitude":42.7019},"firstUpdated":"2016-04-13T12:00:00.000Z","lastUpdated":"2016-10-13T12:00:00.000Z","parameters":["o3"],"countsByMeasurement":[{"parameter":"o3","count":4182}],"count":4182},{"id":"US-1990","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Nashua - Gilson Rd.","locations":["Nashua - Gilson Rd."],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.52242,"latitude":42.71865},"firstUpdated":"2017-02-22T19:00:00.000Z","lastUpdated":"2019-10-09T08:00:00.000Z","parameters":["o3"],"countsByMeasurement":[{"parameter":"o3","count":10536}],"count":10536}]};