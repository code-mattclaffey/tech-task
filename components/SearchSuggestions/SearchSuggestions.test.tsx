import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { SearchSuggestions } from ".";
import { App } from "../App";
import { SearchContextProvider } from '../SearchContext';

afterEach(() => {
    cleanup();
});

describe("<SearchSuggestions />", () => {
    it("should match snapshot", () => {
        const { container } = render(
            <SearchContextProvider>
                <SearchSuggestions />
            </SearchContextProvider>
        );

        expect(container).toMatchSnapshot();
    });

    it("should remember the search term when the component re-renders", () => {
        const expectedResult = "Manchester";
        const { queryByTestId } = render(
            <SearchContextProvider>
                <SearchSuggestions />
            </SearchContextProvider>
        );

        const searchResultInput = queryByTestId("search-result-input");

        if (searchResultInput === null) {
            expect(true).toEqual(false);
            return;
        }

        fireEvent.change(searchResultInput, {
            target: { value: expectedResult }
        });

        expect(searchResultInput.getAttribute("value")).toEqual(expectedResult);
    });
});
