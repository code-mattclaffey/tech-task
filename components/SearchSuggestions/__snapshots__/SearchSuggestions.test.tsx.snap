// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchSuggestions /> should match snapshot 1`] = `
<div>
  <form
    action=""
    autocomplete="off"
    class="c-search-suggestions"
    data-testid="search-form"
  >
    <div
      class="c-input__container c-input__container--has-icon c-input__container--thin"
    >
      <svg
        fill="none"
        height="30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <line
          x1="21"
          x2="16.65"
          y1="21"
          y2="16.65"
        />
      </svg>
      <input
        class="c-input"
        data-testid="search-result-input"
        id="search-term"
        name="search-term"
        placeholder="Enter a city name"
        type="text"
        value=""
      />
      <label
        class="c-label u-visually-hidden"
        for="search-term"
      >
        Enter a city name
      </label>
    </div>
    <div
      aria-live="polite"
      class="u-visually-hidden"
      data-testid="search-results-aria-live"
      role="status"
    >
      0
    </div>
    <div
      class="c-search-suggestions__results"
      data-testid="search-results"
      id="search-results"
      role="listbox"
      tabindex="-1"
    />
  </form>
</div>
`;
